### PR TITLE
Migrating to netstandard + FAKE 5

### PR DESCRIPTION
**Pull request by legezam**
This pull request aims to migrate Amyris.AmyrisBio to dotnet standard and FAKE 5. I used Dmx.AmyrisBio as a base but i did the migration from scratch to make sure everything is okay.

Please note, in order to build this branch, you may need to remove `bin` and `obj` directories and sometimes i even had to clear nuget local caches (`dotnet nuget locals all --clear`)

## Breaking changes

During the migration it turned out that some functionality cannot be migrated directly anymore:
- SourceLink support - as per my understanding, it became a dotnet cli tool and is no longer usable from fake. If it is necessary to have, we need to find a workaround - removed the generator task completely for now
- Generation of documentation - the new version of FSharp.Formatting that supports dotnetstandard uses a different `generate.fsx` that doesn't rely on fakelib. As i found no documentation in the repository, i removed the doc generation completely for now. If we have a decision to make it work again, i am happy to migrate to the new FSharp.Formatting
- The octokit.fsx is not supported in the new fake, however there is a new module, called GitHub which is mostly the same. As the original octokit based release task had the actual file upload commented out, it was unfunctional. Therefore i did a simple 1-1 api conversion in the task and never actually tested.

## Breaking changes in API

Removed windows registry modifier functions as registry is not supported on netstandard:
https://github.com/Amyris/AmyrisBio/pull/4/files#diff-3a791f56821248b5a46100871b0cc6f6

## Recommended process of Amyris.GslCore update

As discussed with @daz10000 , Amyris may want AmyrisBio library to be migrated in parallel with GslCore.

I would imagine this update process for the two libraries:
1. migration of Amyris.GslCore [this PR](https://github.com/Amyris/GslCore/pull/28)
2. upgrading AmyrisBio to dotnetstandard + FAKE5 (This PR)
3. release AmyrisBio
4. i am submitting another PR to GslCore where i change Dmx.AmyrisBio -> AmyrisBio in Amyris.GslCore
5. release GslCore

### CODE REVIEW

**Comment by legezam** [comment]
Please note, if you still need to access windows registry from netcore, you can use [windows compatibility pack](https://www.nuget.org/packages/Microsoft.Windows.Compatibility) as an intermediary solution.

**Comment by chrismacklin** [comment]
Is it possible to build this library in a configuration that supports both netstandard and projects that are still running on .NET framework ~4.6?  The problem with this PR as it stands is that to accept it, I must upgrade every one of our production F# applications together to dotnetcore, which is a large undertaking.

**Comment by daz10000** [comment]
On Tue, May 21, 2019 at 08:38:04PM +0000, Chris Macklin wrote:
If we target netstandard,  I believe it will still support 4.6 (as long as you aren't using platform specific features like accessing registry).  I have taken out some crazy legacy features that relied on the windows api

https://docs.microsoft.com/en-us/dotnet/standard/library-guidance/cross-platform-targeting

Darren

**Comment by chrismacklin** [comment]
Ah, thanks for the clarification.  I was having some issues getting things to build on top of this but with this insight I can see now that I'm looking at red herrings.  Once I can confirm that everything still builds on top of this port, then I can merge this and release a new version.

Why did you need to port this library to support the GSL compiler running on dotnetcore?

**Comment by daz10000** [comment]
> Ah, thanks for the clarification.  I was having some issues getting things to build on top of this but with this insight I can see now that I'm looking at red herrings.  Once I can confirm that everything still builds on top of this port, then I can merge this and release a new version.

Happy to help with any errors if you want to share.

> 
> Why did you need to port this library to support the GSL compiler running on dotnetcore?

Not strictly necessary for any one feature but we moved to netcore over a year ago and it greatly simplifies all the project files / tooling / cross platform support - also made it easier to set up CI.  It makes sending in more PRs for other features we haven't yet submitted a lot simpler.  It's worth it just for the new project files.  You also get warnings (runtime) if you use old style projects with netcore so better just to migrate everything

Darren

**Comment by chrismacklin** [comment]
Got it, thanks for clarifying!  That significantly eases getting this merged.

File: tests/AmyrisBio.Tests/testDnaops.fs [hunk 1]
```diff
@@ -5,7 +5,7 @@ open Amyris.Bio.dnaops
 open Amyris.Bio.biolib
 open Amyris.Bio.utils
 
-let refTestGenome = @"..\..\test_data\S288C.fsa" |> Amyris.Bio.utils.smashSlash
+let refTestGenome = @"..\..\..\test_data\S288C.fsa" |> Amyris.Bio.utils.smashSlash
```
Comment (Yzupnick) [inline tests/AmyrisBio.Tests/testDnaops.fs:8]: This test is passing for me, but do you know why this change had to happen?

**Comment by Yzupnick** [comment]
@legezam, 
Thanks for porting this too. 

I was able to get all the dotnet standard versions working with our internal applications (on .NET 4.6).  However I ran into some issues when our applications used Amyris.Bio while GslCore used Dmx.Amyris.Bio

## Alternate upgrade ordering

   1. Upgrade AmyrisBio to dotnetstandard + FAKE5 (This PR)
   1. Release AmyrisBio
   1. Upgrade GslCore (include reference to AmyrisBio in the existing PR) 
   1. Release GslCore

What are your thoughts?

**Review by Yzupnick — CHANGES_REQUESTED** [review]
Looks great so far. 

I think this PR has gotten a little stale though. Would you mind rebasing on master? (The only conflict should be the Release Notes and Assembly.Fs) 

Thanks!

**Comment by daz10000** [comment]
On Thu, Aug 22, 2019 at 01:09:48PM -0700, Chanan Zupnick wrote:
> Yzupnick requested changes on this pull request.
> 
> Looks great so far. 
> 
> I think this PR has gotten a little stale though. Would you mind rebasing on master? (The only conflict should be the Release Notes and Assembly.Fs) 
> 
> Thanks!
> 
> > @@ -5,7 +5,7 @@ open Amyris.Bio.dnaops
>  open Amyris.Bio.biolib
>  open Amyris.Bio.utils
>  
> -let refTestGenome = @"..\..\test_data\S288C.fsa" |> Amyris.Bio.utils.smashSlash
> +let refTestGenome = @"..\..\..\test_data\S288C.fsa" |> Amyris.Bio.utils.smashSlash
> 
> This test is passing for me, but do you know why this change had to happen?

Yes! - the netcore binary folder structure adds one more layer to the output, so instead of bin/Debug/Foo.dll you get bin/Debug/{architecture}/Foo.dll so the test data are one more level of redirect away.  In some implementations I have made the tests flexible enough to check both when searching to preserve backwards compatibility

**Comment by legezam** [comment]
> 
> 
> Looks great so far.
> 
> I think this PR has gotten a little stale though. Would you mind rebasing on master? (The only conflict should be the Release Notes and Assembly.Fs)
> 
> Thanks!

Done! I have tested the build even with the latest .net core 3 prev 7 sdk. Should be fine to merge now

**Comment by legezam** [comment]
> 
> 
> @legezam,
> Thanks for porting this too.
> 
> I was able to get all the dotnet standard versions working with our internal applications (on .NET 4.6). However I ran into some issues when our applications used Amyris.Bio while GslCore used Dmx.Amyris.Bio
> ## Alternate upgrade ordering
> 
>     1. Upgrade AmyrisBio to dotnetstandard + FAKE5 (This PR)
> 
>     2. Release AmyrisBio
> 
>     3. Upgrade GslCore (include reference to AmyrisBio in the existing PR)
> 
>     4. Release GslCore
> 
> 
> What are your thoughts?

I would prefer this too.

**Comment by legezam** [comment]
@Yzupnick , i have added further upgrades:
- migrated to netstandard2.1 + netcoreapp3.1
- removed manual handling of tool restoration and introduced tool manifest
- updated dependencies

please review

File: src/AmyrisBio/utils.fs [hunk 2]
```diff
@@ -142,7 +142,7 @@ module utils =
     /// Returns files in dir ending with one of a list of suffices
     let filesEndingWith inDir suffixList =
         Directory.GetFiles(inDir)
-        |> Seq.filter (fun name -> List.exists (fun suffix -> name.EndsWith(suffix)) suffixList)
+        |> Seq.filter (fun name -> List.exists (fun (suffix : string) -> name.EndsWith(suffix)) suffixList)
```
Comment (legezam) [inline src/AmyrisBio/utils.fs:145]: please note, string.EndsWith received new overloads in netstandard2.1 making the compiler complain about ambiguity. This type annotation resolves that ambiguity

**Comment by Yzupnick** [comment]
We are ready internally for this!

Thanks for your patience. I am going to do one final review today and if everything looks good/still works I will publish the latest Nuget tomorrow.
I will also be bumping the major version of the Nuget package to 3.0. Does that work for you?

**Comment by daz10000** [comment]
Congrats and yes on version bump. Thanks!  We should celebrate 

Von meinem iPhone gesendet

**Pull request merged**